homeassistant/components/lifx/coordinator.py

```diff
@@ -186,60 +190,85 @@ def async_get_entity_id(self, platform: Platform, key: str) -> str | None:
-                )
-
-            # device.mac_addr is not the mac_address, its the serial number
-            if self.device.mac_addr == TARGET_ANY:
-                self.device.mac_addr = response.target_addr
-
-            if self._update_rssi is True:
-                await self.async_update_rssi()
-
-            # Update extended multizone devices
-            if lifx_features(self.device)["extended_multizone"]:
-                await self.async_get_extended_color_zones()
-                await self.async_get_multizone_effect()
-            # use legacy methods for older devices
-            elif lifx_features(self.device)["multizone"]:
-                await self.async_get_color_zones()
-                await self.async_get_multizone_effect()
+    async def _async_populate_device_info(self) -> None:
+        """Populate device info."""
+        methods: list[Callable] = []
+        device = self.device
+        if self.device.host_firmware_version is None:
+            methods.append(device.get_hostfirmware)
+        if self.device.product is None:
+            methods.append(device.get_version)
+        if self.device.group is None:
+            methods.append(device.get_group)
+        assert methods, "Device info already populated"
+        await async_multi_execute_lifx_with_retries(
+            methods, DEFAULT_ATTEMPTS, OVERALL_TIMEOUT
+        )
 
-            if lifx_features(self.device)["hev"]:
-                await self.async_get_hev_cycle()
+    @callback
+    def _async_build_color_zones_update_requests(self) -> list[Callable]:
+        """Build a color zones update request."""
+        device = self.device
+        return [
+            partial(device.get_color_zones, start_index=zone)
+            for zone in range(0, len(device.color_zones), 8)
+        ]
 
-            if lifx_features(self.device)["infrared"]:
-                await async_execute_lifx(self.device.get_infrared)
+    async def _async_update_data(self) -> None:
+        """Fetch all device data from the api."""
+        device = self.device
+        if (
+            device.host_firmware_version is None
+            or device.product is None
+            or device.group is None
+        ):
+            await self._async_populate_device_info()
+
+        num_zones = len(device.color_zones) if device.color_zones is not None else 0
+        features = lifx_features(self.device)
+        is_extended_multizone = features["extended_multizone"]
+        is_legacy_multizone = not is_extended_multizone and features["multizone"]
+        update_rssi = self._update_rssi
+        methods: list[Callable] = [self.device.get_color]
+        if update_rssi:
+            methods.append(self.device.get_wifiinfo)
+        if is_extended_multizone:
+            methods.append(self.device.get_extended_color_zones)
+            methods.append(self.device.get_multizone_effect)
+        elif is_legacy_multizone:
+            methods.append(self.device.get_extended_color_zones)
```

Comment:
Legacy devices don't support the `get_extended_color_zones` request. This should probably be `get_multizone_effect`.

Reply:
refactoring error. fixed